yoyodyne/models/modules/lstm.py

```diff
@@ -223,3 +223,167 @@ def forward(
+                shape (B x seq_len).
+
+        Returns:
+            output (base.ModuleOutput): Dataclass containing step-wise
+                emission probs, alignment matrix, and hidden states of decoder.
+        """
+        # Encodes current symbol.
+        embedded = self.embed(symbol)
+        decoded, hiddens = self.module(embedded, last_hiddens)
+        # Gets emission probabilities over each hidden state (source symbol).
+        output = decoded.expand(-1, encoder_out.shape[1], -1)
+        output = torch.cat([output, encoder_out], dim=-1)
+        output = self.output_proj(output)
+        # Gets transition probabilities (alignment) for current states.
+        alignment = self._alignment_step(decoded, encoder_out, encoder_mask)
+        return base.ModuleOutput(
+            output=output, embeddings=alignment, hiddens=hiddens
+        )
+
+    def get_module(self) -> nn.LSTM:
+        return nn.LSTM(
+            self.embedding_size,
+            self.hidden_size,
+            num_layers=self.layers,
+            dropout=self.dropout,
+            batch_first=True,
+            bidirectional=self.bidirectional,
+        )
+
+    @property
+    def output_size(self) -> int:
+        return self.decoder_input_size + self.hidden_size
+
+    @property
+    def name(self) -> str:
+        return "HMM LSTM"
+
+
+class ContextHMMLSTMDecoder(HMMLSTMDecoder):
+    def __init__(self, *args, attention_context, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.delta = attention_context
+        self.alignment_proj = nn.Linear(
+            self.hidden_size * 2, (attention_context * 2) + 1
+        )  # Window size must include center and both sides.
+
+    def _alignment_step(self, decoded, encoder_out, encoder_mask):
+        # -> B x seq_len
+        alignment_scores = torch.cat(
+            [self.scale_encoded(encoder_out), decoded], dim=2
+        )
+        alignment_scores = self.alignment_proj(alignment_scores)
+        alignment_probs = nn.functional.softmax(alignment_scores, dim=-1)
+        # Limits context to window of self.delta (context length).
+        alignment_probs = alignment_probs.split(1, dim=1)
+        alignment_probs = torch.cat(
+            [
+                nn.functional.pad(
+                    t,
+                    (
+                        -self.delta + i,
+                        encoder_mask.shape[1] - (self.delta + 1) - i,
+                    ),
+                )
+                for i, t in enumerate(alignment_probs)
+            ],
+            dim=1,
+        )
+        # Get probability of alignments.
+        # Mask padding.
+        alignment_probs = alignment_probs * (~encoder_mask).unsqueeze(1) + 1e-7
+        alignment_probs = alignment_probs / alignment_probs.sum(
+            dim=-1, keepdim=True
+        )
+        return alignment_probs.log()  # Log probs for quicker computation
+
+    def forward(
```

Comment:
I think this can just inherit from the other HMM class no?

Reply:
not quite, you need the expanded decoded for two operations. lemme see if i can reduce boilerplate.